File: backend/src/routes/routes/disputes.js
const express = require('express');
const router = express.Router();
const { PrismaClient } = require('@prisma/client');
const prisma = new PrismaClient();
const { authenticate } = require('../middleware/auth');
const { body, validationResult } = require('express-validator');
const { createNotification } = require('../services/notificationService');

// Open a dispute
router.post(
  '/',
  authenticate,
  [
    body('orderId').notEmpty().withMessage('Order ID is required'),
    body('reason').notEmpty().withMessage('Reason is required'),
  ],
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) return res.status(400).json({ errors: errors.array() });

      const { orderId, reason } = req.body;

      const order = await prisma.order.findUnique({
        where: { id: orderId },
        include: { buyer: true, seller: true },
      });

      if (!order) return res.status(404).json({ error: 'Order not found' });
      if (order.buyerId !== req.user.id && order.sellerId !== req.user.id) {
        return res.status(403).json({ error: 'Not authorized' });
      }
      if (!['PAID', 'SHIPPED', 'DELIVERED'].includes(order.status)) {
        return res.status(400).json({ error: 'Cannot dispute this order' });
      }

      const existing = await prisma.dispute.findFirst({
        where: { orderId, status: { in: ['OPEN', 'UNDER_REVIEW'] } },
      });
      if (existing) return res.status(400).json({ error: 'Active dispute already exists for this order' });

      const dispute = await prisma.dispute.create({
        data: {
          orderId,
          raisedById: req.user.id,
          reason,
          status: 'OPEN',
        },
      });

      await prisma.order.update({
        where: { id: orderId },
        data: { status: 'DISPUTED' },
      });

      // Notify the other party
      const otherUserId = order.buyerId === req.user.id ? order.sellerId : order.buyerId;
      await createNotification({
        userId: otherUserId,
        type: 'DISPUTE_OPENED',
        message: `A dispute has been opened for order #${order.orderNumber}`,
        auctionId: order.auctionId,
      });

      res.status(201).json(dispute);
    } catch (err) {
      next(err);
    }
  }
);

// Get disputes for user
router.get('/', authenticate, async (req, res, next) => {
  try {
    const disputes = await prisma.dispute.findMany({
      where: {
        OR: [
          { raisedById: req.user.id },
          { order: { buyerId: req.user.id } },
          { order: { sellerId: req.user.id } },
        ],
      },
      include: {
        order: {
          include: {
            auction: { select: { id: true, title: true, images: true } },
            buyer: { select: { id: true, username: true } },
            seller: { select: { id: true, username: true } },
          },
        },
        raisedBy: { select: { id: true, username: true } },
      },
      orderBy: { createdAt: 'desc' },
    });
    res.json(disputes);
  } catch (err) {
    next(err);
  }
});

// Get single dispute
router.get('/:id', authenticate, async (req, res, next) => {
  try {
    const dispute = await prisma.dispute.findUnique({
      where: { id: req.params.id },
      include: {
        order: {
          include: {
            auction: true,
            buyer: { select: { id: true, username: true, email: true } },
            seller: { select: { id: true, username: true, email: true } },
          },
        },
        raisedBy: { select: { id: true, username: true } },
      },
    });
    if (!dispute) return res.status(404).json({ error: 'Dispute not found' });

    // Check authorization
    const order = dispute.order;
    if (order.buyerId !== req.user.id && order.sellerId !== req.user.id && dispute.raisedById !== req.user.id) {
      return res.status(403).json({ error: 'Not authorized' });
    }

    res.json(dispute);
  } catch (err) {
    next(err);
  }
});

// Resolve dispute (mock admin action — in production, this would be admin-only)
router.put(
  '/:id/resolve',
  authenticate,
  [body('resolution').notEmpty().withMessage('Resolution is required')],
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) return res.status(400).json({ errors: errors.array() });

      const dispute = await prisma.dispute.findUnique({
        where: { id: req.params.id },
        include: { order: true },
      });

      if (!dispute) return res.status(404).json({ error: 'Dispute not found' });
      if (!['OPEN', 'UNDER_REVIEW'].includes(dispute.status)) {
        return res.status(400).json({ error: 'Dispute is already resolved' });
      }

      const updated = await prisma.dispute.update({
        where: { id: req.params.id },
        data: {
          status: 'RESOLVED',
          resolution: req.body.resolution,
          resolvedAt: new Date(),
        },
      });

      // Update order status back
      await prisma.order.update({
        where: { id: dispute.orderId },
        data: { status: 'COMPLETED' },
      });

      // Notify both parties
      const order = dispute.order;
      for (const userId of [order.buyerId, order.sellerId]) {
        await createNotification({
          userId,
          type: 'DISPUTE_OPENED', // Reusing type for resolution notification
          message: `Dispute for order #${order.orderNumber} has been resolved`,
          auctionId: order.auctionId,
        });
      }

      res.json(updated);
    } catch (err) {
      next(err);
    }
  }
);

module.exports = router;

File: backend/app/routes/disputes.py
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import or_, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.config.database import get_db
from app.middleware.auth import get_current_user
from app.models.models import Dispute, DisputeStatus, Order, OrderStatus, User
from app.services.notification_service import create_notification

router = APIRouter(prefix="/api/disputes", tags=["disputes"])


class OpenDisputeRequest(BaseModel):
    orderId: str
    reason: str


class ResolveDisputeRequest(BaseModel):
    resolution: str


@router.post("", status_code=201)
async def open_dispute(
    body: OpenDisputeRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Order).where(Order.id == body.orderId).options(selectinload(Order.buyer), selectinload(Order.seller))
    )
    order = result.scalar_one_or_none()
    if not order:
        raise HTTPException(status_code=404, detail="Order not found")
    if order.buyerId != user.id and order.sellerId != user.id:
        raise HTTPException(status_code=403, detail="Not authorized")

    order_status = order.status.value if hasattr(order.status, "value") else order.status
    if order_status not in ("PAID", "SHIPPED", "DELIVERED"):
        raise HTTPException(status_code=400, detail="Cannot dispute this order")

    existing_r = await db.execute(
        select(Dispute).where(
            Dispute.orderId == body.orderId,
            Dispute.status.in_([DisputeStatus.OPEN, DisputeStatus.UNDER_REVIEW]),
        )
    )
    if existing_r.scalar_one_or_none():
        raise HTTPException(status_code=400, detail="Active dispute already exists for this order")

    dispute = Dispute(
        orderId=body.orderId,
        raisedById=user.id,
        reason=body.reason,
        status=DisputeStatus.OPEN,
    )
    db.add(dispute)
    await db.commit()
    await db.refresh(dispute)

    other_user_id = order.sellerId if order.buyerId == user.id else order.buyerId
    await create_notification(
        db, other_user_id, "DISPUTE_OPENED",
        "Dispute opened",
        f"A dispute has been opened for order #{order.orderNumber}",
        {"auctionId": order.auctionId},
    )

    return {
        "id": dispute.id,
        "orderId": dispute.orderId,
        "raisedById": dispute.raisedById,
        "reason": dispute.reason,
        "status": dispute.status.value if hasattr(dispute.status, "value") else dispute.status,
        "createdAt": dispute.createdAt.isoformat(),
    }


@router.get("")
async def get_disputes(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Dispute)
        .where(
            or_(
                Dispute.raisedById == user.id,
                Dispute.orderId.in_(select(Order.id).where(Order.buyerId == user.id)),
                Dispute.orderId.in_(select(Order.id).where(Order.sellerId == user.id)),
            )
        )
        .options(
            selectinload(Dispute.order).selectinload(Order.auction),
            selectinload(Dispute.order).selectinload(Order.buyer),
            selectinload(Dispute.order).selectinload(Order.seller),
            selectinload(Dispute.raisedBy),
        )
        .order_by(Dispute.createdAt.desc())
    )
    disputes = result.scalars().all()

    return [
        {
            "id": d.id,
            "reason": d.reason,
            "status": d.status.value if hasattr(d.status, "value") else d.status,
            "resolution": d.resolution,
            "createdAt": d.createdAt.isoformat(),
            "order": {
                "id": d.order.id,
                "orderNumber": d.order.orderNumber,
                "auction": {
                    "id": d.order.auction.id,
                    "title": d.order.auction.title,
                    "images": d.order.auction.images or [],
                } if d.order.auction else None,
                "buyer": {"id": d.order.buyer.id, "username": d.order.buyer.username} if d.order.buyer else None,
                "seller": {"id": d.order.seller.id, "username": d.order.seller.username} if d.order.seller else None,
            } if d.order else None,
            "raisedBy": {"id": d.raisedBy.id, "username": d.raisedBy.username} if d.raisedBy else None,
        }
        for d in disputes
    ]


@router.get("/{dispute_id}")
async def get_dispute(
    dispute_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Dispute)
        .where(Dispute.id == dispute_id)
        .options(
            selectinload(Dispute.order).selectinload(Order.auction),
            selectinload(Dispute.order).selectinload(Order.buyer),
            selectinload(Dispute.order).selectinload(Order.seller),
            selectinload(Dispute.raisedBy),
        )
    )
    dispute = result.scalar_one_or_none()
    if not dispute:
        raise HTTPException(status_code=404, detail="Dispute not found")

    order = dispute.order
    if order.buyerId != user.id and order.sellerId != user.id and dispute.raisedById != user.id:
        raise HTTPException(status_code=403, detail="Not authorized")

    return {
        "id": dispute.id,
        "reason": dispute.reason,
        "description": dispute.description,
        "status": dispute.status.value if hasattr(dispute.status, "value") else dispute.status,
        "resolution": dispute.resolution,
        "resolvedAt": dispute.resolvedAt.isoformat() if dispute.resolvedAt else None,
        "createdAt": dispute.createdAt.isoformat(),
        "order": {
            "id": order.id,
            "orderNumber": order.orderNumber,
            "auction": {
                "id": order.auction.id, "title": order.auction.title,
                "images": order.auction.images or [], "description": order.auction.description,
            } if order.auction else None,
            "buyer": {
                "id": order.buyer.id, "username": order.buyer.username, "email": order.buyer.email,
            } if order.buyer else None,
            "seller": {
                "id": order.seller.id, "username": order.seller.username, "email": order.seller.email,
            } if order.seller else None,
        } if order else None,
        "raisedBy": {"id": dispute.raisedBy.id, "username": dispute.raisedBy.username} if dispute.raisedBy else None,
    }


@router.put("/{dispute_id}/resolve")
async def resolve_dispute(
    dispute_id: str,
    body: ResolveDisputeRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Dispute).where(Dispute.id == dispute_id).options(selectinload(Dispute.order))
    )
    dispute = result.scalar_one_or_none()
    if not dispute:
        raise HTTPException(status_code=404, detail="Dispute not found")

    dispute_status = dispute.status.value if hasattr(dispute.status, "value") else dispute.status
    if dispute_status not in ("OPEN", "UNDER_REVIEW"):
        raise HTTPException(status_code=400, detail="Dispute is already resolved")

    dispute.status = DisputeStatus.RESOLVED
    dispute.resolution = body.resolution
    dispute.resolvedAt = datetime.utcnow()
    dispute.updatedAt = datetime.utcnow()
    await db.commit()

    order = dispute.order
    if order:
        order.status = OrderStatus.COMPLETED
        order.updatedAt = datetime.utcnow()
        await db.commit()

        for uid in [order.buyerId, order.sellerId]:
            await create_notification(
                db, uid, "DISPUTE_OPENED",
                "Dispute resolved",
                f"Dispute for order #{order.orderNumber} has been resolved",
                {"auctionId": order.auctionId},
            )

    await db.refresh(dispute)
    return {
        "id": dispute.id,
        "reason": dispute.reason,
        "status": dispute.status.value if hasattr(dispute.status, "value") else dispute.status,
        "resolution": dispute.resolution,
        "resolvedAt": dispute.resolvedAt.isoformat() if dispute.resolvedAt else None,
        "createdAt": dispute.createdAt.isoformat(),
    }
